attestation/slsa/slsa.go

```diff
@@ -0,0 +1,241 @@
+		switch name := attestor.Attestor.Name(); name {
+		// Pre-material Attestors
+		case environment.Name:
+			envs := attestor.Attestor.(*environment.Attestor).Variables
+			pbEnvs := make(map[string]interface{}, len(envs))
+			for name, value := range envs {
+				pbEnvs[name] = value
+			}
+
+			internalParameters["env"] = pbEnvs
+
+		case git.Name:
+			digestSet := attestor.Attestor.(git.GitAttestor).Data().CommitDigest
+			remotes := attestor.Attestor.(git.GitAttestor).Data().Remotes
+			digests, _ := digestSet.ToNameMap()
+
+			for _, remote := range remotes {
+				p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+					p.PbProvenance.BuildDefinition.ResolvedDependencies,
+					&v1.ResourceDescriptor{
+						Name:   remote,
+						Digest: digests,
+					})
+			}
+
+		case github.Name:
+			gh := attestor.Attestor.(github.GitHubAttestor)
+			p.PbProvenance.RunDetails.Builder.Id = GHABuilderId
+			p.PbProvenance.RunDetails.Metadata.InvocationId = gh.Data().PipelineUrl
+			digest := make(map[string]string)
+			digest["sha1"] = gh.Data().JWT.Claims["sha"].(string)
+
+			p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+				p.PbProvenance.BuildDefinition.ResolvedDependencies,
+				&v1.ResourceDescriptor{
+					Name:   gh.Data().ProjectUrl,
+					Digest: digest,
+				})
+
+		case gitlab.Name:
+			gl := attestor.Attestor.(*gitlab.Attestor)
+			p.PbProvenance.RunDetails.Builder.Id = GLCBuilderId
+			p.PbProvenance.RunDetails.Metadata.InvocationId = gl.PipelineUrl
+			digest := make(map[string]string)
+			digest["sha1"] = gl.JWT.Claims["sha"].(string)
+
+			p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+				p.PbProvenance.BuildDefinition.ResolvedDependencies,
+				&v1.ResourceDescriptor{
+					Name:   gl.ProjectUrl,
+					Digest: digest,
+				})
+
+		// Material Attestors
+		case material.Name:
+			mats := attestor.Attestor.(material.MaterialAttestor).Materials()
+			for name, digestSet := range mats {
+				digests, _ := digestSet.ToNameMap()
+				p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+					p.PbProvenance.BuildDefinition.ResolvedDependencies,
+					&v1.ResourceDescriptor{
+						Name:   name,
+						Digest: digests,
+					})
+			}
+
+		// CommandRun Attestors
+		case commandrun.Name:
+			var err error
+			ep := make(map[string]interface{})
+			ep["command"] = strings.Join(attestor.Attestor.(commandrun.CommandRunAttestor).Data().Cmd, " ")
+			p.PbProvenance.BuildDefinition.ExternalParameters, err = structpb.NewStruct(ep)
+			if err != nil {
+				return err
+			}
+			// We have start and finish time at the collection level, how do we access it here?
+			p.PbProvenance.RunDetails.Metadata.StartedOn = timestamppb.New(time.Now())
+			p.PbProvenance.RunDetails.Metadata.FinishedOn = timestamppb.New(time.Now())
+
+		// Product Attestors
+		case product.ProductName:
+			maps.Copy(p.products, attestor.Attestor.(product.ProductAttestor).Products())
```

Comment:
hey @jkjell ! I pulled this down to see if I can make some QOL improvements, and I noticed that I was getting an `assignment to entry in nil map` error. Is this code running for you? I had to modify the code to initialise `p.products` in order to get it to work :)

Reply:
also would it make sense to leverage ctx.Products() here?